File: 06-DecisionesAnidadas/04-Tarifa/01-Python/Tarifa.py
"""
    Crear un programa para calcular, de forma clara y transparente, 
    el valor de un trayecto en taxi considerando variables como 
    el horario diurno/nocturno, los recargos de domingos y festivos, 
    y el destino (ya sea dentro del perímetro urbano o fuera de él)
    para determinar si la tarifa cobrada es justa o no.
    Autor: Julián Esteban Gutiérrez Posada
    Fecha: Julio 2025
    Licencia: GNU GPL v3
"""

from modulo.util import mostrar_mensaje, ingresar_entero, \
    ingresar_opcion, ingresar_logico

URBANO = 1
PERIFERIA = 2
EXTRA_URBANO = 3
VIA_AEROPUERTO = 4

SIN_COSTO = 0
VALOR_MINIMO = 6000
VALOR_ESPECIAL = 1300

VALOR_PERIFERIA = 3400
VALOR_EXTRA_URBANO = 4900
VALOR_VIA_AEROPUERTO = 1600

def main():
    valor_cobrado = ingresar_entero("Valor cobrado: ")
    valor_taximetro = ingresar_entero("Valor del taxímetro: ")
    es_especial = ingresar_logico("Es domingo, festivo o nocturno (s/n): ")
    destino_trayecto = ingresar_opcion("Destino \n" +
                                       "\t(1) urbano,\n" +
                                       "\t(2) lugar periférico,\n" +
                                       "\t(3) extra urbano\n" +
                                       "\t(4) vía al aeropuerto\n\n" +
                                       "Cuál es su destino: ", 4)

    tarifa_minima  = determinar_tarifa_minima(valor_taximetro)
    tarifa_especial= determinar_tarifa_especial(es_especial)
    tarifa_destino = determinar_tarifa_destino(destino_trayecto) 

    tarifa_real = calcular_tarifa(
        tarifa_minima, tarifa_especial, tarifa_destino)
    
    mensaje_cobro = determinar_mensaje_cobro(valor_cobrado, tarifa_real)

    informe_cobro = generar_informe_cobro(
        valor_cobrado, tarifa_real, mensaje_cobro)

    mostrar_mensaje(informe_cobro)

def determinar_tarifa_minima(valor_taximetro):
    return max(valor_taximetro, VALOR_MINIMO)

def determinar_tarifa_especial(es_especial):
    tarifa_especial = SIN_COSTO

    if es_especial:
        tarifa_especial = VALOR_ESPECIAL
        
    return tarifa_especial

def determinar_tarifa_destino(destino_trayecto):
    tarifa_destino = SIN_COSTO

    if destino_trayecto == PERIFERIA:
        tarifa_destino = VALOR_PERIFERIA
    elif destino_trayecto == EXTRA_URBANO:
        tarifa_destino = VALOR_EXTRA_URBANO
    elif destino_trayecto == VIA_AEROPUERTO:
        tarifa_destino = VALOR_VIA_AEROPUERTO

    return tarifa_destino

def calcular_tarifa(tarifa_minima, tarifa_especial, tarifa_destino):
    return tarifa_minima + tarifa_especial + tarifa_destino

def determinar_mensaje_cobro(valor_cobrado, tarifa_real):
    mensaje_cobro = "JUSTO"

    if valor_cobrado != tarifa_real:
        mensaje_cobro = "INJUSTO"

    return mensaje_cobro

def generar_informe_cobro(valor_cobrado, tarifa_real, mensaje_cobro):
    return (
        f"\nValor cobrado \t${valor_cobrado:7d}"
        f"\nValor real \t${tarifa_real:7d}\n"
        f"\nPor lo anterior el cobro es {mensaje_cobro}\n"
    )

main()

File: 06-DecisionesAnidadas/04-Tarifa/05-Haskell/Tarifa.hs
{-
   Crear un programa para calcular, de forma clara y transparente,
   el valor de un trayecto en taxi considerando variables como
   el horario diurno/nocturno, los recargos de domingos y festivos,
   y el destino (ya sea dentro del perímetro urbano o fuera de él)
   para determinar si la tarifa cobrada es justa o no.
   Autor: Julián Esteban Gutiérrez Posada
   Fecha: Julio 2025
   Licencia: GNU GPL v3
-}

import Text.Printf (printf)
import Modulo.Util (mostrarMensaje, ingresarEntero, ingresarOpcion, ingresarLogico)

urbano, periferia, extraUrbano, viaAeropuerto :: Int
urbano        = 1
periferia     = 2
extraUrbano   = 3
viaAeropuerto = 4

sinCosto, valorMinimo, valorEspecial :: Int
sinCosto      = 0
valorMinimo   = 6000
valorEspecial = 1300

valorPeriferia, valorExtraUrbano, valorViaAeropuerto :: Int
valorPeriferia     = 3400
valorExtraUrbano   = 4900
valorViaAeropuerto = 1600

main :: IO()
main = do
  valorCobrado <- ingresarEntero "Valor cobrado: "
  valorTaximetro <- ingresarEntero "Valor del taxímetro: "
  esEspecial <- ingresarLogico "Es domingo, festivo o nocturno (s/n): "
  destinoTrayecto <- ingresarOpcion ( "Destino \n" ++
    "\t(1) urbano,\n" ++
    "\t(2) lugar periférico,\n" ++
    "\t(3) extra urbano\n" ++
    "\t(4) vía al aeropuerto\n\n" ++
    "Cuál es su destino: " ) 4
  
  let tarifaMinima = determinarTarifaMinima valorTaximetro
  let tarifaEspecial = determinarTarifaEspecial esEspecial
  let tarifaDestino = determinarTarifaDestino destinoTrayecto

  let tarifaReal = calcularTarifa tarifaMinima tarifaEspecial tarifaDestino

  let mensajeCobro = determinarMensajeCobro valorCobrado tarifaReal
  
  let informeCobro = generarInformeCobro valorCobrado tarifaReal mensajeCobro
  
  mostrarMensaje informeCobro

determinarTarifaMinima :: Int -> Int
determinarTarifaMinima valorTaximetro = 
  max valorTaximetro valorMinimo

determinarTarifaEspecial :: Bool -> Int
determinarTarifaEspecial esEspecial
  | esEspecial = valorEspecial
  | otherwise  = sinCosto

determinarTarifaDestino :: Int -> Int
determinarTarifaDestino destinoTrayecto
  | destinoTrayecto == periferia     = valorPeriferia
  | destinoTrayecto == extraUrbano   = valorExtraUrbano
  | destinoTrayecto == viaAeropuerto = valorViaAeropuerto
  | otherwise                        = sinCosto

calcularTarifa :: Int -> Int -> Int -> Int
calcularTarifa tarifaMinima tarifaEspecial tarifaDestino =
  tarifaMinima + tarifaEspecial + tarifaDestino

determinarMensajeCobro ::  Int -> Int -> String
determinarMensajeCobro valorCobrado tarifaReal
  | valorCobrado == tarifaReal = "JUSTO"
  | otherwise                  = "INJUSTO"

generarInformeCobro :: Int -> Int -> String -> String
generarInformeCobro valorCobrado tarifaReal mensajeCobro =
  printf ("\nValor cobrado \t$%7d" ++
    "\nValor real   \t$%7d"++
    "\n\nPor lo anterior el cobro es %s\n" )
    valorCobrado tarifaReal mensajeCobro